src/tess_base_api.rs

```diff
@@ -86,6 +88,44 @@ pub struct TessBaseApiGetWordStrBoxTextError();
 #[error("TessBaseApi get_component_images returned null")]
 pub struct TessBaseApiGetComponentImagesError();
 
+#[derive(Debug, Error)]
+#[error("TessBaseApi all_word_confidences returned null")]
+pub struct TessBaseApiAllWordConfidencesError();
+
+pub struct AllWordConfidences(*mut c_int, usize);
+
+impl AllWordConfidences {
+    pub fn as_slice(&self) -> &[i32] {
+        self
+    }
+
+    pub fn as_slice_mut(&mut self) -> &mut [i32] {
+        self
+    }
+}
+
+impl Deref for AllWordConfidences {
+    type Target = [i32];
```

Comment:
Would it be better to use `c_int` here so that on platforms where `i32` and `c_int` aren't the same this still compiles?

```suggestion
    pub fn as_slice(&self) -> &[c_int] {
        self
    }

    pub fn as_slice_mut(&mut self) -> &mut [c_int] {
        self
    }
}

impl Deref for AllWordConfidences {
    type Target = [c_int];
```

Reply:
Ah I didn't realise you were doing that elsewhere, I'll change it if you wish.

However my personal opinion is to not expose c types in the high level api. I've done the same sort of thing before with char, which is signed on some platforms and unsigned on others. I'd personally just make the return be an i32 and cast the c value to and i32.

Reply:
> I'd personally just make the return be an i32 and cast the c value to and i32.

It's difficult. It would be wrong to cast the pointer to a different type before using it as a slice. And after it's a slice it can't be casted unless you were to allocate a Vec.

I've been trying to have the higher level crates (`leptess` and `tesseract-rs`) do the casting to Rust types.

Reply:
Ah true I forgot this case is a slice. Could always make this an iterator instead but you'd sacrifise random access.

> I've been trying to have the higher level crates (leptess and tesseract-rs) do the casting to Rust types.

Don't really get why. a -sys and a high level crate is the usual way to go. Another level of abstraction feels like it would just get in the way to me.

Reply:
> Don't really get why. a -sys and a high level crate is the usual way to go. Another level of abstraction feels like it would just get in the way to me.

It seemed sensible at the time, but it looks like the decision has aged poorly.

There were two crates doing much the same thing (leptess and tesseract-rs), both of which I help maintain. They both took in the `tesseract-sys` library and did some work to make it safe and create an opinionated higher level API.

There were 2 problems with this:
1. There was a lot of duplication between the 2 crates
2. There were a couple instances where memory was freed at the wrong points. I think this was due to the complexity of having the same function responsible for using the c API safely and presenting a high level interface.

I agree having 3 levels of crates is unusual.

What would you do in my situation?

Reply:
Just a normal -sys and higher level crate. The higher level crate should be basically 1 to one with the c in functionality.

If people want to build other libs around the high level then sure but I don't see a point it reimplementing everything.

If it's just some extra functionality the crate could be tesseract-utils or something.